File: pakalon-backend/app/streaming.py
"""Stream optimization for Pakalon backend.

Provides buffered streaming for provider responses.
"""

import asyncio
import logging
from typing import AsyncIterator, Any

logger = logging.getLogger(__name__)


class StreamOptimizer:
    """Optimizes streaming responses with buffering."""

    def __init__(self, buffer_size: int = 10, flush_interval_ms: int = 100):
        self._buffer: list[str] = []
        self._buffer_size: int = buffer_size
        self._flush_interval: float = flush_interval_ms / 1000.0
        self._on_chunk_callback: Any = None

    def set_callback(self, callback) -> None:
        """Set callback for chunk events."""
        self._on_chunk_callback = callback

    async def process_stream(
        self,
        stream: AsyncIterator[str],
        buffer_size: int | None = None,
    ) -> AsyncIterator[str]:
        """Process a stream with buffering."""
        buf_size = buffer_size or self._buffer_size
        buffer: list[str] = []

        async for chunk in stream:
            buffer.append(chunk)

            if len(buffer) >= buf_size:
                combined = "".join(buffer)
                buffer.clear()
                yield combined

        # Flush remaining
        if buffer:
            yield "".join(buffer)

    async def batch_stream(
        self,
        stream: AsyncIterator[str],
        max_batch_size: int = 50,
        timeout_seconds: float = 0.1,
    ) -> AsyncIterator[str]:
        """Batch stream chunks with timeout."""
        batch: list[str] = []
        last_flush = asyncio.get_event_loop().time()

        async for chunk in stream:
            batch.append(chunk)
            current_time = asyncio.get_event_loop().time()

            should_flush = (
                len(batch) >= max_batch_size
                or (current_time - last_flush) >= timeout_seconds
            )

            if should_flush and batch:
                yield "".join(batch)
                batch.clear()
                last_flush = current_time

        if batch:
            yield "".join(batch)


# Global instance
stream_optimizer = StreamOptimizer()

File: pakalon-cli/src/providers/streaming.ts
/**
 * Stream Optimization for Pakalon CLI
 *
 * Provides buffered streaming for provider responses.
 */

import logger from "@/utils/logger.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface StreamConfig {
  bufferSize?: number;
  flushIntervalMs?: number;
}

// ---------------------------------------------------------------------------
// Default Configurations
// ---------------------------------------------------------------------------

const DEFAULT_CONFIG: StreamConfig = {
  bufferSize: 10,
  flushIntervalMs: 100,
};

// ---------------------------------------------------------------------------
// Stream Optimizer
// ---------------------------------------------------------------------------

export class StreamOptimizer {
  private buffer: string[] = [];
  private config: StreamConfig;

  constructor(config: StreamConfig = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /**
   * Process a stream with buffering
   */
  async *processStream(
    stream: AsyncIterable<string>,
    bufferSize?: number
  ): AsyncGenerator<string, void, unknown> {
    const bufSize = bufferSize || this.config.bufferSize || 10;
    const buffer: string[] = [];

    for await (const chunk of stream) {
      buffer.push(chunk);

      if (buffer.length >= bufSize) {
        yield buffer.join("");
        buffer.length = 0;
      }
    }

    // Flush remaining
    if (buffer.length > 0) {
      yield buffer.join("");
    }
  }

  /**
   * Batch stream chunks with timeout
   */
  async *batchStream(
    stream: AsyncIterable<string>,
    maxBatchSize: number = 50,
    timeoutMs: number = 100
  ): AsyncGenerator<string, void, unknown> {
    const batch: string[] = [];
    let lastFlush = Date.now();

    for await (const chunk of stream) {
      batch.push(chunk);
      const now = Date.now();

      const shouldFlush =
        batch.length >= maxBatchSize ||
        now - lastFlush >= timeoutMs;

      if (shouldFlush && batch.length > 0) {
        yield batch.join("");
        batch.length = 0;
        lastFlush = now;
      }
    }

    if (batch.length > 0) {
      yield batch.join("");
    }
  }

  /**
   * Add chunk to buffer
   */
  addChunk(chunk: string): void {
    this.buffer.push(chunk);
  }

  /**
   * Check if buffer should be flushed
   */
  shouldFlush(): boolean {
    return this.buffer.length >= (this.config.bufferSize || 10);
  }

  /**
   * Flush buffer and return combined string
   */
  flush(): string {
    if (this.buffer.length === 0) {
      return "";
    }

    const combined = this.buffer.join("");
    this.buffer = [];
    return combined;
  }

  /**
   * Get current buffer size
   */
  getBufferSize(): number {
    return this.buffer.length;
  }

  /**
   * Clear buffer
   */
  clear(): void {
    this.buffer = [];
  }
}

// Global instance
export const streamOptimizer = new StreamOptimizer();

// ---------------------------------------------------------------------------
// Convenience Functions
// ---------------------------------------------------------------------------

/**
 * Process stream with buffering
 */
export async function* processStream(
  stream: AsyncIterable<string>,
  bufferSize?: number
): AsyncGenerator<string, void, unknown> {
  yield* streamOptimizer.processStream(stream, bufferSize);
}

/**
 * Batch stream chunks
 */
export async function* batchStream(
  stream: AsyncIterable<string>,
  maxBatchSize?: number,
  timeoutMs?: number
): AsyncGenerator<string, void, unknown> {
  yield* streamOptimizer.batchStream(stream, maxBatchSize, timeoutMs);
}
